pkg/capabilities/validate.go

```diff
@@ -0,0 +1,210 @@
+		ValidatorArgs: args,
+		schemas:       make(map[string]string),
+	}
+}
+
+func (v *Validator[Config, Input, Output]) Schema() (string, error) {
+	type Combined struct {
+		Config Config `json:"config"`
+		Input  Input  `json:"input"`
+		Output Output `json:"output"`
+	}
+	combined := Combined{}
+
+	type CombinedWithoutInput struct {
+		Config Config `json:"config"`
+		Output Output `json:"output"`
+	}
+	var combinedWithoutInput = CombinedWithoutInput{}
+
+	var config interface{} = combined.Config
+	var input interface{} = combined.Input
+	var output interface{} = combined.Output
+	if config == nil {
+		return "", errors.New("config is nil, please provide a config type")
+	}
+	if output == nil {
+		return "", errors.New("output is nil, please provide an output type")
+	}
+
+	// we allow input to be nil, since triggers do not have inputs
+	if input == nil {
+		return schemaWith(v.SchemaReflector, combinedWithoutInput, v.schemas, "root", v.Info)
+	}
+
+	// print values of combined
+	return schemaWith(v.SchemaReflector, combined, v.schemas, "root", v.Info)
+}
+func (v *Validator[Config, Input, Output]) ConfigSchema() (string, error) {
+	return schemaWith(v.ConfigReflector, v.configType, v.schemas, "config", v.Info)
+}
+func (v *Validator[Config, Input, Output]) InputSchema() (string, error) {
+	return schemaWith(v.InputReflector, v.inputType, v.schemas, "input", v.Info)
+}
+func (v *Validator[Config, Input, Output]) OutputSchema() (string, error) {
+	return schemaWith(v.OutputReflector, v.outputType, v.schemas, "output", v.Info)
+}
+
+func (v *Validator[Config, Input, Output]) ValidateInput(input *values.Map) (*Input, error) {
+	inputSchema, err := v.InputSchema()
+	if err != nil {
+		return nil, errors.Join(errors.New("validation error while validating input"), err)
+	}
+
+	return validateAgainstSchema[Input](input, inputSchema)
+}
+func (v *Validator[Config, Input, Output]) ValidateConfig(config *values.Map) (*Config, error) {
+	configSchema, err := v.ConfigSchema()
+	if err != nil {
+		return nil, errors.Join(errors.New("validation error while validating config"), err)
+	}
+
+	return validateAgainstSchema[Config](config, configSchema)
+}
+func (v *Validator[Config, Input, Output]) ValidateOutput(output *values.Map) (*Output, error) {
+	outputSchema, err := v.OutputSchema()
+	if err != nil {
+		return nil, errors.Join(errors.New("validation error while validating output"), err)
+	}
+
+	return validateAgainstSchema[Output](output, outputSchema)
+}
+
+func validateAgainstSchema[DecodedValue any](value *values.Map, schema string) (*DecodedValue, error) {
+	jsonSchema, err := jsonvalidate.CompileString(uriPrefix, schema)
+	if err != nil {
+		return nil, err
+	}
+
+	raw, err := value.Unwrap()
+	if err != nil {
+		return nil, err
+	}
+
+	err = jsonSchema.Validate(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	var decodedValue *DecodedValue
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
```

Comment:
If we're decoding `value` into a struct here, we should try to use UnwrapTo if we can -- same thing with the custom decode hook: if we're not handling an edge case correctly, let's put it in the values library rather than here.

Reply:
Sure, could you explain why to me? I'm not knowledgable on when I should choose using `UnwrapTo` rather than just using mapstructure directly